### DISCUSSION TITLE
Problems with the props of a react-table v8 alpha HOC

### DISCUSSION THREAD

**Discussion by Obiwarn** [TanStack/table#3784] [Q&A]
I am testing `@tanstack/react-table: 8.0.0-alpha.11`, which works great so far.

Now, I try to create a reusable Table component for my app that encapsulates most of react-table.

I have a hard time to get the typings right, could you give me a hint @tannerlinsley?

Also, I am not sure if I understand the API right but: it seems that I have to pass all three props? (since `table` is needed to create `columns` outside the component and needed in the `useTable` hook inside the component?).

So far I got this:

```
interface Props<T> {
  table: TableFactory<T>;
  data: T[];
  columns: ColumnDef<Partial<DefaultGenerics>>[];
}

export const SuperTable = <T extends object>(props:Props<T>) => {
    // ...
```

**Comment by tannerlinsley** [comment]
TableFactory takes partial generics too

https://github.com/TanStack/react-table/blob/f10003b398be44d2f85e94874b7c8d307b464754/packages/table-core/src/createTable.tsx#L46

Tanner Linsley

**Reply by Obiwarn** [reply]
Thank you!

For the second question: Is there any way I can avoid that the user of the HOC has to create and pass the table object separately?


Final result is:
```
interface Props<T> {
  table: TableFactory<Overwrite<Partial<DefaultGenerics>, { Row: T }>>;
  data: T[];
  columns: ColumnDef<Overwrite<Partial<DefaultGenerics>, { Row: T }>>[];
}
```

**Reply by knoefel** [reply]
@Obiwarn Great question, i'm facing the same problem right now. Have you been able to solve it?

**Reply by knoefel** [reply]
@tannerlinsley I'm facing a similar problem right now trying to get the typings right. So in my use case i have a re-usable table component, which is used by a parent component. The idea is to setup the columns in the parent and pass both the return value of `createTable` and the columns to the custom table component.

Here is my solution so far, the problem is, i'm losing some type information (e.g. when using `getRowId`):

**Parent**
```typescript
export default function Assets() {
  const table = useMemo(() => createTable().setRowType<ListAssetFragment>(), []);
  const columns = useMemo(
    () => [  // column definitions  ],
    []
  );

  return (
    <>
      <Table table={table} columns={columns} data={data} rowId="id" />
    </>
  );
}
```

**Table**
```typescript
type TableProps<T, G> = {
  table: ReactTable<Overwrite<G, { Row: T }>>;
  columns: ColumnDef<Overwrite<G, { Row: T }>>[];
  data: T[];
  rowId: keyof T;
};

export const Table = <T extends object, G extends ReactTableGenerics>({
  table,
  data,
  pageCount,
  rowId,
  columns,
}: TableProps<T, G>) => {
  const instance = useTableInstance(table, {
    data,
    columns,
    getRowId: (row) => row[rowId], // here i lose my type bindings
  });

  // render method...
}
```
I'm a little stuck right now on how to solve this. I would be awesome to get some feedback or tips on how to handle this. Thx in advance!!

**Reply by edwin177** [reply]
@knoefel Your example seems to work for me. I just had to do `String(row[rowId])` in `getRowId` since some of my id's are integers.

**Reply by talatkuyuk** [reply]
**Which one is better react-table experience for creating custom Table Component ?**
1. (First Approach) Create table + columns; then pass **the table + the columns + the data** to the custom table component. And create the table instance in the custom table component 
2. (Second Approach) Create table + columns + instance; then pass **the instance + the data** to the custom table component. _(I am currently using this)_

**Reply by knoefel** [reply]
@talatkuyuk i think the first approach is handling additional configuration better, e.g. adding pagination as you can do the setup in your custom table component and therefore encapsulate this logic. Choosing the second approach you would need to pass/setup the pagination state to the table `instance` for every component where you want to use the table component.

**Reply by knoefel** [reply]
> @knoefel Your example seems to work for me. I just had to do `String(row[rowId])` in `getRowId` since some of my id's are integers.

@edwin177 That's interesting. In my case `row` is losing the type bindings, as shown here:

<img width="517" alt="Screenshot 2022-05-23 at 12 09 07" src="https://user-images.githubusercontent.com/2396344/169797202-63452f68-3113-4868-b0fd-a33e8b849727.png">

In an earlier version where i don't have implemented the generics typing yet, i could only access properties on `row` which were indeed keys of `T` so it would pass for `rowId`, which i typed as `keyof T`, but would fail for any arbitrary value.

**Reply by edwin177** [reply]
@knoefel This is my setup so far:

![image](https://user-images.githubusercontent.com/8732768/170154480-ce0a3886-e651-45d4-aaa0-d6d3dc18cae9.png)

**Reply by knoefel** [reply]
@edwin177 That's great. I just tried it out. I like it, it's much cleaner than my version. Thanks a lot for sharing!!
